### PR TITLE
Plugin Support | Users can use custom plugins with Delivery SDK

### PR DESCRIPTION
**Pull request by nadeem-cs**
Plugin support for JS Delivery SDK added

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -94,6 +95,17 @@ export interface FetchOptions {
...
 
+//Plugins
+export interface ContentstackPlugin {
+    onRequest?(stack: Stack, request: ContentstackRequest): ContentstackRequest;
+    onResponse?(stack: Stack, request: ContentstackRequest, response: any, data: any): any;
+}
+
+export interface ContentstackRequest {
+    url: string;
+    headers: object;
```
Comment (uttamukkoji) [inline index.d.ts:106]: change headers to `option` as per implementation here: https://github.com/contentstack/contentstack-javascript/blob/33df49f790ecb277d3d260a1ab9fa33731019397/src/core/lib/request.js#L101

File: src/core/lib/request.js [hunk 2]
```diff
@@ -83,20 +85,44 @@ function fetchRetry(url, headers, fetchOptions, resolve, reject, retryDelay = 30
...
+        for (let index = 0; index < plugins.length; index++) {
+
+            if (typeof plugins[index].onRequest == 'function') {
+                request = plugins[index].onRequest(stack, request)
+            }
+        }
+    }
+
+    
     fetch(url, option)
```
Comment (uttamukkoji) [inline src/core/lib/request.js:111]: @nadeem-cs please update url and option with request object.

File: src/core/stack.js [hunk 3]
```diff
@@ -88,6 +97,7 @@ export default class Stack {
                     };
                     if (typeof stack_arguments[0].live_preview == "object") {
                         this.live_preview =  Utils.mergeDeep(this.config.live_preview , stack_arguments[0].live_preview)
+                        console.log(this.live_preview);
```
Comment (uttamukkoji) [inline src/core/stack.js:100]: @nadeem-cs please remove `console.log`

File: src/core/lib/request.js [hunk 4]
```diff
@@ -83,20 +85,45 @@ function fetchRetry(url, headers, fetchOptions, resolve, reject, retryDelay = 30
...
                 })
         }
     }
+
     if (fetchOptions.debug)  fetchOptions.logHandler('info', { url: url, option: option});
-    fetch(url, option)
-        .then(function(response) {    
+
+    let request = {url, option};
+
```
Comment (uttamukkoji) [inline src/core/lib/request.js:99]: @nadeem-cs make variable `plugins` to `let`

File: src/core/lib/request.js [hunk 5]
```diff
@@ -108,7 +109,7 @@ function fetchRetry(stack, queryParams, fetchOptions, resolve, reject, retryDela
     }
 
     
-    fetch(url, option)
+    fetch(request)
```
Comment (uttamukkoji) [inline src/core/lib/request.js:112]: @nadeem-cs  it should be fetch(request.url, request.option)

**Review by uttamukkoji — APPROVED** [review]

**Review by ishaileshmishra — APPROVED** [review]
approved

File: src/core/lib/request.js [hunk 6]
```diff
@@ -58,7 +57,10 @@ function wait(retryDelay) {
     });
 }
 
-function fetchRetry(url, headers, fetchOptions, resolve, reject, retryDelay = 300, retryLimit = 5) {
+function fetchRetry(stack, queryParams, fetchOptions, resolve, reject, retryDelay = 300, retryLimit = 5) {
+    let requestParams = stack.requestParams,
```
Comment (renuka0reddy) [inline src/core/lib/request.js:61]: Are these being reassigned, if not let's use const?
Consider removing var from the below line too

File: src/core/lib/request.js [hunk 7]
```diff
@@ -83,20 +85,45 @@ function fetchRetry(url, headers, fetchOptions, resolve, reject, retryDelay = 30
...
     }
+
     if (fetchOptions.debug)  fetchOptions.logHandler('info', { url: url, option: option});
-    fetch(url, option)
-        .then(function(response) {    
+
+    let request = {url, option};
+
+    let plugins = stack.plugins;
+    if (plugins && plugins != undefined) {
```
Comment (renuka0reddy) [inline src/core/lib/request.js:101]: != => !==

File: src/core/lib/request.js [hunk 8]
```diff
@@ -83,20 +85,45 @@ function fetchRetry(url, headers, fetchOptions, resolve, reject, retryDelay = 30
...
+
+        for (let index = 0; index < plugins.length; index++) {
+
+            if (typeof plugins[index].onRequest == 'function') {
```
Comment (renuka0reddy) [inline src/core/lib/request.js:105]: Always use typed checked, === instead of ==

File: src/core/lib/request.js [hunk 9]
```diff
@@ -83,20 +85,45 @@ function fetchRetry(url, headers, fetchOptions, resolve, reject, retryDelay = 30
...
+
+    
+    fetch(request.url, request.option)
+        .then( function(response) {
+            
             if (fetchOptions.debug)  fetchOptions.logHandler('info', response);
-            let data = response.json();      
+            let data = response.json();
+
             if (response.ok && response.status === 200) {
```
Comment (renuka0reddy) [inline src/core/lib/request.js:118]: this entire block of code including the older one, looks pretty complex wiht too many if else, can we simplify it and make more readable?

File: src/core/lib/request.js [hunk 10]
```diff
@@ -83,20 +85,45 @@ function fetchRetry(url, headers, fetchOptions, resolve, reject, retryDelay = 30
...
+        }
+    }
+
+    
+    fetch(request.url, request.option)
+        .then( function(response) {
+            
             if (fetchOptions.debug)  fetchOptions.logHandler('info', response);
-            let data = response.json();      
+            let data = response.json();
```
Comment (renuka0reddy) [inline src/core/lib/request.js:116]: const?

File: src/core/lib/request.js [hunk 11]
```diff
@@ -83,20 +85,45 @@ function fetchRetry(url, headers, fetchOptions, resolve, reject, retryDelay = 30
...
+
             if (response.ok && response.status === 200) {
-                resolve(data);
+                data.then(json => {
+                    for (let index = 0; index < plugins.length; index++) {
+                        if (typeof plugins[index].onResponse == 'function')
```
Comment (renuka0reddy) [inline src/core/lib/request.js:121]: === here and everywhere else

File: src/core/lib/request.js [hunk 12]
```diff
@@ -83,20 +85,45 @@ function fetchRetry(url, headers, fetchOptions, resolve, reject, retryDelay = 30
...
+
     if (fetchOptions.debug)  fetchOptions.logHandler('info', { url: url, option: option});
-    fetch(url, option)
-        .then(function(response) {    
+
+    let request = {url, option};
+
+    let plugins = stack.plugins;
+    if (plugins && plugins != undefined) {
+
```
Comment (renuka0reddy) [inline src/core/lib/request.js:102]: Quite some additions here, no tests changing or added?

**Pull request merged**